diag_manager/fms_diag_file_object.F90

```diff
@@ -0,0 +1,407 @@
+  integer, dimension(:), allocatable :: var_ids !< Variable IDs corresponding to var_list
+  integer, dimension(:), private, allocatable :: var_index !< An array of the variable indicies in the 
+                                                                 !! diag_object.  This should be the same size as
+                                                                 !! `file_varlist`
+  logical, dimension(:), private, allocatable :: var_reg   !< Array corresponding to `file_varlist`, .true. 
+                                                                 !! if the variable has been registered and 
+                                                                 !! `file_var_index` has been set for the variable
+
+ contains
+
+  procedure, public :: has_file_metadata_from_model
+  procedure, public :: has_fileobj
+#ifdef use_yaml
+  procedure, public :: has_diag_yaml
+#endif
+  procedure, public :: has_var_ids
+  procedure, public :: has_var_list
+  procedure, public :: get_id
+! TODO  procedure, public :: get_fileobj ! TODO
+  procedure, public :: get_file_domain_type
+! TODO  procedure, public :: get_diag_yaml ! TODO
+  procedure, public :: get_file_metadata_from_model
+  procedure, public :: get_var_list
+  procedure, public :: get_var_ids
+! The following fuctions come will use the yaml inquiry functions
+#ifdef use_yaml
+ procedure :: get_file_fname
+ procedure :: get_file_frequnit
+ procedure :: get_file_freq
+ procedure :: get_file_timeunit
+ procedure :: get_file_unlimdim
+!! TODO get functions for sub region stuff
+! procedure :: get_file_sub_region
+ procedure :: get_file_new_file_freq
+ procedure :: get_file_new_file_freq_units
+ procedure :: get_file_start_time
+ procedure :: get_file_duration
+ procedure :: get_file_duration_units
+ procedure :: get_file_varlist
+ procedure :: get_file_global_meta
+ procedure :: has_file_fname 
+ procedure :: has_file_frequnit 
+ procedure :: has_file_freq 
+ procedure :: has_file_timeunit 
+ procedure :: has_file_unlimdim 
+ procedure :: has_file_sub_region 
+ procedure :: has_file_new_file_freq 
+ procedure :: has_file_new_file_freq_units 
+ procedure :: has_file_start_time 
+ procedure :: has_file_duration 
+ procedure :: has_file_duration_units 
+ procedure :: has_file_varlist 
+ procedure :: has_file_global_meta 
+#endif
+
+end type fmsDiagFile_type
+
+type(fmsDiagFile_type), dimension (:), allocatable, target :: FMS_diag_files !< The array of diag files
+
+contains
+
+!< @brief Allocates the number of files and sets an ID based for each file
+!! @return true if there are files allocated in the YAML object
+logical function fms_diag_files_object_init ()
+ integer :: nFiles !< Number of files in the diag yaml
+ integer :: i !< Looping iterator
+ if (diag_yaml%has_diag_files()) then
+        nFiles = diag_yaml%size_diag_files()
+        allocate (FMS_diag_files(nFiles))
+        set_ids: do i=1,nFiles
+                FMS_diag_files(i)%diag_yaml => diag_yaml
```

Comment:
this should really be diag_yaml%diag_files(i), you only need the yaml info for the one file not the whole thing